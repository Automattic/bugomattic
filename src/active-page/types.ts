export type ActivePage = 'search-for-issues' | 'report-issue';
