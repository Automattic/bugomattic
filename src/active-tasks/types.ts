import { Task } from '../reporting-config';

export type EntityType = 'product' | 'featureGroup' | 'feature';

export interface SourcedTask {
	sourceId: string;
	sourceType: EntityType;
	details: Task;
}

export interface ActiveTask extends SourcedTask {
	completed: boolean;
}

export type TaskIndex = number;
