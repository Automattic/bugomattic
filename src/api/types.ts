import { LearnMoreLink, TaskMapping } from '../reporting-config';

/**
 * An interface representing an interaction with the API.
 * Can create multiple implementations to support local testing and production deployment.
 */
export interface ApiClient {
	loadReportingConfig(): Promise< ReportingConfigApiResponse >;
}

/**
 * The issue reporting config as stored in and fetched from the database.
 * It follows a more heirarchical structure for easy reading and manual updating.
 */
export interface ReportingConfigApiResponse {
	[ productName: string ]: Product;
}

interface Product {
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	featureGroups?: FeatureGroups;
	features?: Features;
}

interface FeatureGroups {
	[ featureGroupName: string ]: FeatureGroup;
}

interface FeatureGroup {
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	features: Features;
}

interface Features {
	[ featureName: string ]: Feature;
}

interface Feature {
	keywords?: string[];
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
}
