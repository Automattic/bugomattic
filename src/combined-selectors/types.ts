export type MatchType = 'name' | 'keyword' | 'description' | 'child';
export type MatchOption = NameMatch | KeywordMatch | DescriptionMatch | ChildMatch;

interface ChildMatch {
	matchType: 'child';
}
interface NameMatch {
	matchType: 'name';
}

interface KeywordMatch {
	matchType: 'keyword';
	keyword: string;
}

interface DescriptionMatch {
	matchType: 'description';
	matchedTerms: Set< string >;
}

interface Matches {
	[ entityId: string ]: MatchOption;
}

export interface ReportingConfigSearchResults {
	products: Matches;
	featureGroups: Matches;
	features: Matches;
}
