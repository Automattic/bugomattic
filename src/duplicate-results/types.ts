export type IssueStatus = 'open' | 'closed';

export interface Issue {
	dateCreated: string;
	dateUpdated: string;
	title: string;
	url: string;
	repo: string;
	status: IssueStatus;
	author: string;
	content: string;
}

export interface DuplicateResultsState {
	results: Issue[];
	requestStatus: 'pending' | 'fulfilled' | 'error';
	requestError: string | null;
}
