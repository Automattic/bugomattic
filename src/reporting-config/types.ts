export interface ReportingConfig {
	normalized: NormalizedReportingConfig;
	indexed: IndexedReportingConfig;
}

export interface NormalizedReportingConfig {
	features: Features;
	featureGroups: FeatureGroups;
	products: Products;
}

export interface IndexedReportingConfig {
	foo: string; // To-do, figure out what indices we need.
}

interface Features {
	[ featureId: string ]: Feature;
}

interface Feature {
	id: string;
	name: string;
	keywords?: string[];
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	featureGroup?: string;
	product?: string;
}

interface FeatureGroups {
	[ featureGroupId: string ]: FeatureGroup;
}

interface FeatureGroup {
	id: string;
	name: string;
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
	product: string;
}

interface Products {
	[ productId: string ]: Product;
}

interface Product {
	id: string;
	name: string;
	description?: string;
	learnMoreLinks?: LearnMoreLink[];
	taskMapping?: TaskMapping;
}

export interface SlackLink {
	type: 'slack';
	channel: string;
}

export interface P2Link {
	type: 'p2';
	subdomain: string;
}

export interface GeneralLink {
	type: 'general';
	displayText?: string;
	href: string;
}

export interface NewGitHubIssueLink {
	type: 'github';
	repository: string; // Automattic/wp-calypso, e.g.
	template?: string;
	projectSlugs?: string[];
	labels?: string[];
}

export type LearnMoreLink = SlackLink | P2Link | GeneralLink;

export type TaskLink = LearnMoreLink | NewGitHubIssueLink;

export interface Task {
	instructions?: string;
	link?: TaskLink;
}

export interface LearnMore {
	description?: string;
	links?: LearnMoreLink[];
}

export interface TaskMapping {
	bug: Task[];
	featureRequest: Task[];
	showStopper: Task[];
}
