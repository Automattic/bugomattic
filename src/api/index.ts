export * from './types';
export * from './local';
export * from './use';
