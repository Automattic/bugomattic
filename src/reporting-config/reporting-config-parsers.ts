import { ApiFeatureGroups, ApiFeatures, ApiTasks, ReportingConfigApiResponse } from '../api/types';
import {
	FeatureGroups,
	FeatureParentEntityType,
	Features,
	IndexedReportingConfig,
	NormalizedReportingConfig,
	Products,
	TaskMapping,
	TaskParentEntityType,
	Tasks,
} from './types';

// Defining all the core logic for how we normalize and index reporting configs here.
// We are keeping this separate from how this logic is used in reducers and thunks.
// This allows easy unit testing of this critical logic apart from how its consumed.

export function normalizeReportingConfig(
	apiResponse: ReportingConfigApiResponse
): NormalizedReportingConfig {
	let normalizedTasks: Tasks = {};
	let normalizedFeatures: Features = {};
	let normalizedFeatureGroups: FeatureGroups = {};
	const normalizedProducts: Products = {};

	for ( const productName in apiResponse ) {
		const productDetails = apiResponse[ productName ];
		if ( ! productDetails || typeof productDetails !== 'object' ) {
			throw new Error(
				`Invalid reporting config. Product "${ productName }" is not a valid product.`
			);
		}

		const { featureGroups, features, description, learnMoreLinks, tasks } =
			apiResponse[ productName ];
		const productId = productName; // We can use name as ID, as being top level, they are unique.

		let productTaskMapping: TaskMapping | undefined;
		if ( tasks ) {
			const taskOutput = normalizeTasks( tasks, {
				parentType: 'product',
				parentId: productId,
			} );
			normalizedTasks = {
				...normalizedTasks,
				...taskOutput.normalizedTasks,
			};
			productTaskMapping = taskOutput.taskMapping;
		}

		let featureGroupIds: string[] = [];
		if ( featureGroups ) {
			const featureGroupOutput = normalizeFeatureGroups( featureGroups, { productId: productId } );
			normalizedTasks = {
				...normalizedTasks,
				...featureGroupOutput.normalizedTasks,
			};
			normalizedFeatures = {
				...normalizedFeatures,
				...featureGroupOutput.normalizedFeatures,
			};
			normalizedFeatureGroups = {
				...normalizedFeatureGroups,
				...featureGroupOutput.normalizedFeatureGroups,
			};

			featureGroupIds = Object.keys( featureGroupOutput.normalizedFeatureGroups );
		}

		let featureIds: string[] = [];
		if ( features ) {
			const featureOutput = normalizeFeatures( features, {
				parentType: 'product',
				parentId: productId,
			} );
			normalizedTasks = {
				...normalizedTasks,
				...featureOutput.normalizedTasks,
			};
			normalizedFeatures = {
				...normalizedFeatures,
				...featureOutput.normalizedFeatures,
			};

			featureIds = Object.keys( featureOutput.normalizedFeatures );
		}

		normalizedProducts[ productId ] = {
			id: productId,
			name: productName,
			description: description,
			learnMoreLinks: learnMoreLinks,
			taskMapping: productTaskMapping,
			featureGroupIds: featureGroupIds,
			featureIds: featureIds,
		};
	}

	return {
		tasks: normalizedTasks,
		features: normalizedFeatures,
		featureGroups: normalizedFeatureGroups,
		products: normalizedProducts,
	};
}

interface FeatureGroupContext {
	productId: string;
}

interface NormalizedFeatureGroupOutput {
	normalizedTasks: Tasks;
	normalizedFeatures: Features;
	normalizedFeatureGroups: FeatureGroups;
}

function normalizeFeatureGroups(
	apiFeatureGroups: ApiFeatureGroups,
	context: FeatureGroupContext
): NormalizedFeatureGroupOutput {
	let normalizedTasks: Tasks = {};
	let normalizedFeatures: Features = {};
	const normalizedFeatureGroups: FeatureGroups = {};

	for ( const featureGroupName in apiFeatureGroups ) {
		const { features, description, learnMoreLinks, tasks } = apiFeatureGroups[ featureGroupName ];
		const featureGroupId = `${ context.productId }__${ featureGroupName }`;

		let featureGroupTaskMapping: TaskMapping | undefined;

		if ( tasks ) {
			const taskOutput = normalizeTasks( tasks, {
				parentId: featureGroupId,
				parentType: 'featureGroup',
			} );
			normalizedTasks = {
				...normalizedTasks,
				...taskOutput.normalizedTasks,
			};
			featureGroupTaskMapping = taskOutput.taskMapping;
		}

		const featureOutput = normalizeFeatures( features, {
			parentType: 'featureGroup',
			parentId: featureGroupId,
		} );
		normalizedFeatures = {
			...normalizedFeatures,
			...featureOutput.normalizedFeatures,
		};
		normalizedTasks = {
			...normalizedTasks,
			...featureOutput.normalizedTasks,
		};

		const featureIds = Object.keys( featureOutput.normalizedFeatures );

		normalizedFeatureGroups[ featureGroupId ] = {
			id: featureGroupId,
			name: featureGroupName,
			description: description,
			learnMoreLinks: learnMoreLinks,
			taskMapping: featureGroupTaskMapping,
			productId: context.productId,
			featureIds: featureIds,
		};
	}

	return {
		normalizedTasks,
		normalizedFeatures,
		normalizedFeatureGroups,
	};
}

interface FeatureContext {
	parentType: FeatureParentEntityType;
	parentId: string;
}

interface NormalizedFeatureOutput {
	normalizedTasks: Tasks;
	normalizedFeatures: Features;
}

function normalizeFeatures(
	apiFeatures: ApiFeatures,
	context: FeatureContext
): NormalizedFeatureOutput {
	let normalizedTasks: Tasks = {};
	const normalizedFeatures: Features = {};
	for ( const featureName in apiFeatures ) {
		const { description, keywords, learnMoreLinks, tasks } = apiFeatures[ featureName ];
		const featureId = `${ context.parentId }__${ featureName }`;

		let featureTaskMapping: TaskMapping | undefined;

		if ( tasks ) {
			const taskOutput = normalizeTasks( tasks, {
				parentId: featureId,
				parentType: 'feature',
			} );
			normalizedTasks = {
				...normalizedTasks,
				...taskOutput.normalizedTasks,
			};
			featureTaskMapping = taskOutput.taskMapping;
		}

		normalizedFeatures[ featureId ] = {
			id: featureId,
			name: featureName,
			description: description,
			keywords: keywords,
			learnMoreLinks: learnMoreLinks,
			taskMapping: featureTaskMapping,
			parentType: context.parentType,
			parentId: context.parentId,
		};
	}

	return {
		normalizedTasks,
		normalizedFeatures,
	};
}

interface TaskContext {
	parentType: TaskParentEntityType;
	parentId: string;
}

interface NormalizedTaskOutput {
	normalizedTasks: Tasks;
	taskMapping: TaskMapping;
}

function normalizeTasks( apiTasks: ApiTasks, context: TaskContext ): NormalizedTaskOutput {
	const taskMapping: TaskMapping = {
		bug: [],
		urgent: [],
		featureRequest: [],
	};
	const normalizedTasks: Tasks = {};

	const issueTypes = [ 'bug', 'featureRequest', 'urgent' ] as const;
	issueTypes.forEach( ( issueType ) => {
		apiTasks[ issueType ].forEach( ( taskDetails, index ) => {
			const taskId = `${ context.parentId }__${ issueType }__${ index }`;
			normalizedTasks[ taskId ] = {
				...taskDetails,
				id: taskId,
				parentType: context.parentType,
				parentId: context.parentId,
			};
			taskMapping[ issueType ].push( taskId );
		} );
	} );

	return { normalizedTasks, taskMapping };
}

export function indexReportingConfig(
	// eslint-disable-next-line @typescript-eslint/no-unused-vars
	_response: ReportingConfigApiResponse
): IndexedReportingConfig {
	// TODO: create real implementation once we know the indices we need!

	return {
		foo: 'bar',
	};
}
