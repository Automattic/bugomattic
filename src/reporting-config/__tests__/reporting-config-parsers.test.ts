import { ReportingConfigApiResponse } from '../../api/types';
import { normalizeReportingConfig } from '../reporting-config-parsers';

function createFakeReportingConfigResponse(): ReportingConfigApiResponse {
	return {
		FakeProduct: {
			description: 'Description for fake product',
			learnMoreLinks: [
				{
					type: 'slack',
					channel: 'product-channel',
				},
			],
			tasks: {
				bug: [ { title: 'Product | bug' } ],
				featureRequest: [ { title: 'Product | feature-request' } ],
				blocker: [ { title: 'Product | blocker' } ],
			},
			featureGroups: {
				FakeFeatureGroup: {
					description: 'Description for fake feature group',
					learnMoreLinks: [
						{
							type: 'p2',
							subdomain: 'feature-group',
						},
					],
					tasks: {
						bug: [ { title: 'Feature group | bug' } ],
						featureRequest: [ { title: 'Feature group | feature request' } ],
						blocker: [ { title: 'Feature group | blocker' } ],
					},
					features: {
						GroupNestedFeature: {
							description: 'Fake feature nested under feature group',
							keywords: [ 'foo' ],
							learnMoreLinks: [
								{
									type: 'slack',
									channel: 'feature-group-feature',
								},
							],
							tasks: {
								bug: [
									{ title: 'Feature nested under group | bug | index 0' },
									{ title: 'Feature nested under group | bug | index 1' },
								],
								featureRequest: [
									{ title: 'Feature nested under group | feature request | index 0' },
									{ title: 'Feature nested under group | feature request | index 1' },
								],
								blocker: [
									{ title: 'Feature nested under group | blocker | index 0' },
									{ title: 'Feature nested under group | blocker | index 1' },
								],
							},
						},
					},
				},
			},
			features: {
				ProductNestedFeature: {
					description: 'Fake feature nested under product',
					keywords: [ 'bar' ],
					learnMoreLinks: [
						{
							type: 'slack',
							channel: 'product-feature',
						},
					],
					tasks: {
						bug: [ { title: 'Feature nested under product bug task' } ],
						featureRequest: [ { title: 'Feature nested under product feature-request task' } ],
						blocker: [ { title: 'Feature nested under product blocker task' } ],
					},
				},
			},
		},
	};
}

describe( '[reporting-config-parsers]', () => {
	describe( 'Normalizing the Reporting Config', () => {
		test( 'Creates the expected normalized reporting config', () => {
			// For this stage of early development, let's use a snapshot to make sure this keeps giving us
			// the respected normalized output. We can add more granular testing later.
			const fakeResponse = createFakeReportingConfigResponse();
			const normalizedReportingConfig = normalizeReportingConfig( fakeResponse );
			expect( normalizedReportingConfig ).toMatchSnapshot();
		} );
	} );
} );
