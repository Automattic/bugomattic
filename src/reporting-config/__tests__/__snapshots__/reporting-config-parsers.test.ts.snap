// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[reporting-config-parsers] Normalizing the Reporting Config Creates the expected normalized reporting config 1`] = `
Object {
  "featureGroups": Object {
    "FakeProduct__FakeFeatureGroup": Object {
      "description": "Description for fake feature group",
      "featureIds": Array [
        "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      ],
      "id": "FakeProduct__FakeFeatureGroup",
      "learnMoreLinks": Array [
        Object {
          "subdomain": "feature-group",
          "type": "p2",
        },
      ],
      "name": "FakeFeatureGroup",
      "productId": "FakeProduct",
      "taskMapping": Object {
        "blocker": Array [
          "FakeProduct__FakeFeatureGroup__blocker__0",
        ],
        "bug": Array [
          "FakeProduct__FakeFeatureGroup__bug__0",
        ],
        "featureRequest": Array [
          "FakeProduct__FakeFeatureGroup__featureRequest__0",
        ],
      },
    },
  },
  "features": Object {
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature": Object {
      "description": "Fake feature nested under feature group",
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "keywords": Array [
        "foo",
      ],
      "learnMoreLinks": Array [
        Object {
          "channel": "feature-group-feature",
          "type": "slack",
        },
      ],
      "name": "GroupNestedFeature",
      "parentId": "FakeProduct__FakeFeatureGroup",
      "parentType": "featureGroup",
      "taskMapping": Object {
        "blocker": Array [
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__0",
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__1",
        ],
        "bug": Array [
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__0",
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__1",
        ],
        "featureRequest": Array [
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__0",
          "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__1",
        ],
      },
    },
    "FakeProduct__ProductNestedFeature": Object {
      "description": "Fake feature nested under product",
      "id": "FakeProduct__ProductNestedFeature",
      "keywords": Array [
        "bar",
      ],
      "learnMoreLinks": Array [
        Object {
          "channel": "product-feature",
          "type": "slack",
        },
      ],
      "name": "ProductNestedFeature",
      "parentId": "FakeProduct",
      "parentType": "product",
      "taskMapping": Object {
        "blocker": Array [
          "FakeProduct__ProductNestedFeature__blocker__0",
        ],
        "bug": Array [
          "FakeProduct__ProductNestedFeature__bug__0",
        ],
        "featureRequest": Array [
          "FakeProduct__ProductNestedFeature__featureRequest__0",
        ],
      },
    },
  },
  "products": Object {
    "FakeProduct": Object {
      "description": "Description for fake product",
      "featureGroupIds": Array [
        "FakeProduct__FakeFeatureGroup",
      ],
      "featureIds": Array [
        "FakeProduct__ProductNestedFeature",
      ],
      "id": "FakeProduct",
      "learnMoreLinks": Array [
        Object {
          "channel": "product-channel",
          "type": "slack",
        },
      ],
      "name": "FakeProduct",
      "taskMapping": Object {
        "blocker": Array [
          "FakeProduct__blocker__0",
        ],
        "bug": Array [
          "FakeProduct__bug__0",
        ],
        "featureRequest": Array [
          "FakeProduct__featureRequest__0",
        ],
      },
    },
  },
  "tasks": Object {
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__0",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | blocker | index 0",
    },
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__1": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__blocker__1",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | blocker | index 1",
    },
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__0",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | bug | index 0",
    },
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__1": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__bug__1",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | bug | index 1",
    },
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__0",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | feature request | index 0",
    },
    "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__1": Object {
      "id": "FakeProduct__FakeFeatureGroup__GroupNestedFeature__featureRequest__1",
      "parentId": "FakeProduct__FakeFeatureGroup__GroupNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under group | feature request | index 1",
    },
    "FakeProduct__FakeFeatureGroup__blocker__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__blocker__0",
      "parentId": "FakeProduct__FakeFeatureGroup",
      "parentType": "featureGroup",
      "title": "Feature group | blocker",
    },
    "FakeProduct__FakeFeatureGroup__bug__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__bug__0",
      "parentId": "FakeProduct__FakeFeatureGroup",
      "parentType": "featureGroup",
      "title": "Feature group | bug",
    },
    "FakeProduct__FakeFeatureGroup__featureRequest__0": Object {
      "id": "FakeProduct__FakeFeatureGroup__featureRequest__0",
      "parentId": "FakeProduct__FakeFeatureGroup",
      "parentType": "featureGroup",
      "title": "Feature group | feature request",
    },
    "FakeProduct__ProductNestedFeature__blocker__0": Object {
      "id": "FakeProduct__ProductNestedFeature__blocker__0",
      "parentId": "FakeProduct__ProductNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under product blocker task",
    },
    "FakeProduct__ProductNestedFeature__bug__0": Object {
      "id": "FakeProduct__ProductNestedFeature__bug__0",
      "parentId": "FakeProduct__ProductNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under product bug task",
    },
    "FakeProduct__ProductNestedFeature__featureRequest__0": Object {
      "id": "FakeProduct__ProductNestedFeature__featureRequest__0",
      "parentId": "FakeProduct__ProductNestedFeature",
      "parentType": "feature",
      "title": "Feature nested under product feature-request task",
    },
    "FakeProduct__blocker__0": Object {
      "id": "FakeProduct__blocker__0",
      "parentId": "FakeProduct",
      "parentType": "product",
      "title": "Product | blocker",
    },
    "FakeProduct__bug__0": Object {
      "id": "FakeProduct__bug__0",
      "parentId": "FakeProduct",
      "parentType": "product",
      "title": "Product | bug",
    },
    "FakeProduct__featureRequest__0": Object {
      "id": "FakeProduct__featureRequest__0",
      "parentId": "FakeProduct",
      "parentType": "product",
      "title": "Product | feature-request",
    },
  },
}
`;
